modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/QueryMemoryManager.java

```diff
@@ -244,9 +228,20 @@ public boolean onQuotaExceeded(long size) {
      *
      * @param newGlobalQuota New global query quota.
      */
-    public synchronized void setGlobalQuota(String newGlobalQuota) {
-        this.globalQuota = U.parseBytes(newGlobalQuota);
-        this.globalQuotaStr = newGlobalQuota;
+    public void setGlobalQuota(String newGlobalQuota) {
+        long globalQuota0 = U.parseBytes(newGlobalQuota);
+        long heapSize = Runtime.getRuntime().maxMemory();
+
+        A.ensure(
+            heapSize > globalQuota0,
+            "Sql global memory quota can't be more than heap size: heapSize="
+                + heapSize + ", quotaSize=" + globalQuota0
+        );
+
+        A.ensure(globalQuota0 >= 0, "Sql global memory quota must be >= 0: quotaSize=" + globalQuota0);
+
+        globalQuota = globalQuota0;
+        globalQuotaStr = newGlobalQuota;
```

Comment:
non-atomic update
Race is possible while updating  globalQuota and globalQuotaStr from different threads.